translator/trace/jaeger/traces_to_jaegerproto.go

```diff
@@ -0,0 +1,356 @@
+
+			jSpan, err := spanToJaegerProto(span)
+			if err != nil {
+				return nil, err
+			}
+			if jSpan != nil {
+				jSpans = append(jSpans, jSpan)
+			}
+		}
+	}
+
+	batch.Spans = jSpans
+
+	return batch, nil
+}
+
+func resourceToJaegerProtoProcess(resource pdata.Resource) *model.Process {
+	if resource.IsNil() {
+		return nil
+	}
+
+	attrs := resource.Attributes()
+	if attrs.Cap() == 0 {
+		return nil
+	}
+
+	process := model.Process{}
+	if serviceName, ok := attrs.Get(conventions.AttributeServiceName); ok {
+		process.ServiceName = serviceName.StringVal()
+		attrs.Delete(conventions.AttributeServiceName)
+	}
+	process.Tags = attributesToJaegerProtoTags(attrs)
+	return &process
+
+}
+
+func attributesToJaegerProtoTags(attrs pdata.AttributeMap) []model.KeyValue {
+	if attrs.Cap() == 0 {
+		return nil
+	}
+
+	tags := make([]model.KeyValue, 0, attrs.Cap())
+	attrs.ForEach(func(key string, attr pdata.AttributeValue) {
+		tag := model.KeyValue{Key: key}
+		switch attr.Type() {
+		case pdata.AttributeValueSTRING:
+			// Jaeger-to-Internal maps binary tags to string attributes and encodes them as
+			// base64 strings. Blindingly attempting to decode base64 seems too much.
+			tag.VType = model.ValueType_STRING
+			tag.VStr = attr.StringVal()
+		case pdata.AttributeValueINT:
+			tag.VType = model.ValueType_INT64
+			tag.VInt64 = attr.IntVal()
+		case pdata.AttributeValueBOOL:
+			tag.VType = model.ValueType_BOOL
+			tag.VBool = attr.BoolVal()
+		case pdata.AttributeValueDOUBLE:
+			tag.VType = model.ValueType_FLOAT64
+			tag.VFloat64 = attr.DoubleVal()
+		}
+		tags = append(tags, tag)
+	})
+	return tags
+}
+
+func spanToJaegerProto(span pdata.Span) (*model.Span, error) {
+	if span.IsNil() {
+		return nil, nil
+	}
+
+	traceID, err := traceIDToJaegerProto(span.TraceID())
+	if err != nil {
+		return nil, err
+	}
+
+	spanID, err := spanIDToJaegerProto(span.SpanID())
+	if err != nil {
+		return nil, err
+	}
+
+	jReferences, err := makeJaegerProtoReferences(span.Links(), span.ParentSpanID(), traceID)
+	if err != nil {
+		return nil, fmt.Errorf("error converting span links to Jaeger references: %w", err)
+	}
+
+	tags := attributesToJaegerProtoTags(span.Attributes())
+	tags = appendTagFromSpanKind(tags, span.Kind())
+	tags = appendTagFromSpanStatus(tags, span.Status())
+	startTime := internal.UnixNanoToTime(span.StartTime())
+
+	return &model.Span{
+		TraceID:       traceID,
+		SpanID:        spanID,
+		OperationName: span.Name(),
+		References:    jReferences,
+		StartTime:     startTime,
+		Duration:      internal.UnixNanoToTime(span.EndTime()).Sub(startTime),
+		Tags:          tags,
+		Logs:          spanEventsToJaegerProtoLogs(span.Events()),
+	}, nil
+}
+
+func traceIDToJaegerProto(traceID pdata.TraceID) (model.TraceID, error) {
+	traceIDHigh, traceIDLow, err := tracetranslator.BytesToUInt64TraceID([]byte(traceID))
+	if err != nil {
+		return model.TraceID{}, err
+	}
+	if traceIDLow == 0 && traceIDHigh == 0 {
+		return model.TraceID{}, errZeroTraceID
```

Comment:
Not for this PR but we changed the receivers to not drop data based on "bad IDs" we should also not drop them at the export time. /cc @tigrannajaryan

Reply:
I think this is the only thing the can error the conversion this will simplify the code since there will be less need for error handling.

Reply:
@pjanotti that's a good point. I was wondering if the translation functions should not return errors at all and just register erroneous data they encounter in metrics and continue a best effort translation.

Reply:
@tigrannajaryan I agree: no errors from translation functions just metrics and debugging logging (or damped logging). We need to improve `obsreport` to better report those.